### DISCUSSION TITLE
Cannot Update from WSL 1 to WSL 2

### DISCUSSION THREAD

**Discussion by dostarora97** [microsoft/WSL#5922] [General]
# Environment

```
Windows build number: Microsoft Windows [Version 10.0.19041.508]
Your Distribution version: Ubuntu 16.04
Linux version 4.4.0-19041-Microsoft (Microsoft@Microsoft.com) (gcc version 5.4.0 (GCC) ) #488-Microsoft Mon Sep 01 13:43:00 PST 2020
```

# Steps to reproduce

**I need to update from WSL 1 to WSL 2.**   

I have read and followed these instructions : [Windows Subsystem for Linux Installation Guide for Windows 10](https://docs.microsoft.com/en-us/windows/wsl/install-win10)

These commands were run in Powershell(administrator) :  

1. `wsl -l -v`
![image](https://user-images.githubusercontent.com/31973060/93140045-69c78c00-f6ff-11ea-8c16-914c1798b70e.png)

2. `dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart`
![image](https://user-images.githubusercontent.com/31973060/93140387-f96d3a80-f6ff-11ea-85c7-10ed91d31922.png)

3. System Restart...

4. `dism.exe /online /enable-feature /featurename:VirtualMachinePlatform /all /norestart`
![image](https://user-images.githubusercontent.com/31973060/93140594-57018700-f700-11ea-8346-63ffd3c067ce.png)

Virtualization Enabled as seen in Task-Manager:
![image](https://user-images.githubusercontent.com/31973060/93142185-27a04980-f703-11ea-8e6b-f27e99fa076f.png)

5. System Restart...

6. `wsl --set-default-version 2`
![image](https://user-images.githubusercontent.com/31973060/93140694-83b59e80-f700-11ea-98c9-aa89181ec404.png)  
I guess this is not the **Expected output**, because [here](https://docs.microsoft.com/en-us/windows/wsl/install-win10#set-wsl-2-as-your-default-version) it says the output should be : `WSL 2 requires an update to its kernel component. For information please visit https://aka.ms/wsl2kernel`

7. `wsl -l -v`
![image](https://user-images.githubusercontent.com/31973060/93140866-cb3c2a80-f700-11ea-9e7b-cc808a7a2bfd.png)

**WSL logs**: https://aka.ms/AA9olbc

#  Expected behavior

The WSL should update from version 1 to version 2. That is after running command `wsl -l -v` in Powershell(administrator), it should say `2` under `VERSION`

# Actual behavior

Even after following the instructions to update WSL, it's version still remains `1`

**Comment by MathijsNL** [comment]
Did you enable this Windows update option? I believe you need it enabled.

![image](https://user-images.githubusercontent.com/1612886/93148501-392e2500-f6f4-11ea-908f-c1a40db95106.png)

Edit:
And also, what is the output for the following command in PowerShell?
```
wsl cat /proc/version
```

**Comment by dostarora97** [comment]
I have that Windows option **Enabled**

```
> wsl cat /proc/version
Linux version 4.4.0-19041-Microsoft (Microsoft@Microsoft.com) (gcc version 5.4.0 (GCC) ) #488-Microsoft Mon Sep 01 13:43:00 PST 2020
```
![image](https://user-images.githubusercontent.com/31973060/93171137-25abaa00-f746-11ea-9437-7ae16434792c.png)

**Comment by power4** [comment]
I have the same problem.the same win version(Version 10.0.19041.508).i can't upgrade to wsl2 always

**Comment by therealkenc** [comment]
> 6. wsl --set-default-version 2

You want `wsl.exe --set-version Legacy 2`. The `--set-default-version` switch does just that; it changes the default. It doesn't change the version of the currently installed distro(s). The output in the OP is correct. Docs to set your distribution version [here](https://docs.microsoft.com/en-us/windows/wsl/install-win10#set-your-distribution-version-to-wsl-1-or-wsl-2).

As an aside, "Legacy" is not all that common to have around anymore. The `--set-version 2` should take, but all things being equal, installing a new distro from the Store is probably not a bad idea in the circumstance.

**Reply by dostarora97** [reply]
> > 1. wsl --set-default-version 2
> 
> You want `wsl.exe --set-version Legacy 2`. The `--set-default-version` switch does just that; it changes the default. It doesn't change the version of the currently installed distro(s). The output in the OP is correct. Docs to set your distribution version [here](https://docs.microsoft.com/en-us/windows/wsl/install-win10#set-your-distribution-version-to-wsl-1-or-wsl-2).
> 
> As an aside, "Legacy" is not all that common to have around anymore. The `--set-version 2` should take, but all things being equal, installing a new distro from the Store is probably not a bad idea in the circumstance.

Alright thanks @therealkenc, so I had few questions.

1. Does that mean setting the version of WSL to `2` does not mean anything in itself? And the version number should actually be considered a property of a distro rather than that of WSL.(to be more accurate) ?

2. Is it that when `wsl.exe --set-default-version 2` is executed, it registers `2` as the default version for any new distro that will be installed ?

3. Does executing `wsl.exe --set-default-version 2` update the version number for already installed distros?

**Reply by therealkenc** [reply]
> Does that mean setting the version of WSL to 2 does not mean anything in itself?

In itself, it means set the default version. _Default_, contrast the _current_ version of any particular distro.

> And the version number should actually be considered a property of a distro rather than that of WSL.

The version of a distro is a property of a distro, yes. You can have WSL1 and WSL2 distributions installed simultaneously, and change their version after the fact. You can get the version of a particular distro with `wsl -l -v`.

> Is it that when wsl.exe --set-default-version 2 is executed, it registers 2 as the default version for any new distro that will be installed ?

Yes. Installed from the MSFT Store, or `--import`ed.

> Does executing wsl.exe --set-default-version 2 update the version number for already installed distros?

That was answered "It doesn't change the version of the currently installed distro(s)" from the previous post.

**Reply by molxyz** [reply]
I had the same issue and tried `wsl.exe --set-default-version 2` several times but it didnot work.  So I deleted Ubuntu 18-04-LTS and installed Ubuntu 20-04-LTS from Microsoft store and it shows version 2.

**Reply by therealkenc** [reply]
> wsl.exe --set-default-version 2

You need `--set-version` on the existing distro same #6505. Your newly installed distro (Ubuntu-20.04) took on the default.

**Reply by Gunnnn** [reply]
> > wsl.exe --set-default-version 2
> 
> You need `--set-version` on the existing distro same #6505. Your newly installed distro (Ubuntu-20.04) took on the default.

In my similar case it doesn't work eiether. No error message, only dumb links
![Image 4](https://user-images.githubusercontent.com/7666505/133929814-3d1d6c86-aad4-4858-afe6-ff30ac6be903.png)
.

**Reply by rearming** [reply]
DISM /Online /Enable-Feature /All /FeatureName:Microsoft-Hyper-V

**Reply by Norlandz** [reply]
> You want `wsl.exe --set-version Legacy 2`. 

For an visual example, see https://admcpr.com/how-to-upgrade-wsl-1-to-wsl-2/
